File: choco3/Zebra.java
/**
  *
  * Zebra problem in Choco3.
  *
  * """
  * This is the zebra problem as invented by Lewis Caroll.
  *
  * There are five houses.
  * The Englishman lives in the red house.
  * The Spaniard owns the dog.
  * Coffee is drunk in the green house.
  * The Ukrainian drinks tea.
  * The green house is immediately to the right of the ivory house.
  * The Old Gold smoker owns snails.
  * Kools are smoked in the yellow house.
  * Milk is drunk in the middle house.
  * The Norwegian lives in the first house.
  * The man who smokes Chesterfields lives in the house next to the man
  *   with the fox.
  * Kools are smoked in the house next to the house where the horse is kept.
  * The Lucky Strike smoker drinks orange juice.
  * The Japanese smokes Parliaments.
  * The Norwegian lives next to the blue house.
  *
  * Who owns a zebra and who drinks water?
  * """
  * 
  * This Choco3 model was created by Hakan Kjellerstrand (hakank@bonetmail.com)
  * Also, see my Choco page: http://www.hakank.org/choco/ 
  *
  */
import org.kohsuke.args4j.Option;
import org.slf4j.LoggerFactory;
import samples.AbstractProblem;
import solver.ResolutionPolicy;
import solver.Solver;
import solver.constraints.Constraint;
import solver.constraints.IntConstraintFactory;
import solver.constraints.IntConstraintFactory.*;
import solver.search.strategy.IntStrategyFactory;
import solver.variables.IntVar;
import solver.variables.BoolVar;
import solver.variables.VariableFactory;
import util.ESat;
import util.tools.ArrayUtils;

import java.util.*;

public class Zebra extends AbstractProblem {

  int n = 5;

  IntVar[] colors;
  IntVar[] nationality;
  IntVar[] animal;
  IntVar[] drink;
  IntVar[] smoke;
  IntVar[] all;

  @Override
  public void createSolver() {
    solver = new Solver("Zebra");
  }
  

  @Override
  public void buildModel() {    

    // Colors
    IntVar red = VariableFactory.enumerated("red", 1, n, solver);
    IntVar green = VariableFactory.enumerated("green", 1, n, solver);
    IntVar yellow = VariableFactory.enumerated("yellow", 1, n, solver);
    IntVar blue = VariableFactory.enumerated("blue", 1, n, solver);
    IntVar ivory = VariableFactory.enumerated("ivory", 1, n, solver);
    colors = new IntVar[] {red, green, yellow, blue, ivory};

    // Nationality
    IntVar englishman = VariableFactory.enumerated("englishman", 1, n, solver);
    IntVar spaniard = VariableFactory.enumerated("spaniard", 1, n, solver);
    IntVar japanese = VariableFactory.enumerated("japanese", 1, n, solver);
    IntVar ukranian = VariableFactory.enumerated("ukranian", 1, n, solver);    
    IntVar norwegian = VariableFactory.enumerated("norwegian", 1, n, solver);
    nationality = new IntVar[] {englishman, spaniard, japanese, ukranian, norwegian};
    
    // Animal
    IntVar dog = VariableFactory.enumerated("dog", 1, n, solver);
    IntVar snails = VariableFactory.enumerated("snails", 1, n, solver);
    IntVar fox = VariableFactory.enumerated("fox", 1, n, solver);
    IntVar zebra = VariableFactory.enumerated("zebra", 1, n, solver);    
    IntVar horse = VariableFactory.enumerated("horse", 1, n, solver);
    animal = new IntVar[] {dog, snails, fox, zebra, horse};
    
    // Drink
    IntVar tea = VariableFactory.enumerated("tea", 1, n, solver);
    IntVar coffee = VariableFactory.enumerated("coffee", 1, n, solver);
    IntVar water = VariableFactory.enumerated("water", 1, n, solver);
    IntVar milk = VariableFactory.enumerated("milk", 1, n, solver);    
    IntVar fruit_juice = VariableFactory.enumerated("fruit_juice", 1, n, solver);
    drink = new IntVar[] {tea, coffee, water, milk, fruit_juice};
    
    // Smoke
    IntVar old_gold = VariableFactory.enumerated("old_gold", 1, n, solver);
    IntVar kools = VariableFactory.enumerated("kools", 1, n, solver);
    IntVar chesterfields = VariableFactory.enumerated("chesterfields", 1, n, solver);
    IntVar lucky_strike = VariableFactory.enumerated("lucky_strike", 1, n, solver);    
    IntVar parliaments = VariableFactory.enumerated("parliaments", 1, n, solver);
    smoke = new IntVar[] {old_gold, kools, chesterfields, lucky_strike, parliaments};


    solver.post(IntConstraintFactory.alldifferent(colors, "BC"));
    solver.post(IntConstraintFactory.alldifferent(nationality, "BC"));
    solver.post(IntConstraintFactory.alldifferent(animal, "BC"));
    solver.post(IntConstraintFactory.alldifferent(drink, "BC"));
    solver.post(IntConstraintFactory.alldifferent(smoke, "BC"));

    all = new IntVar[n*5];
    all = ArrayUtils.append(colors, nationality, animal, drink, smoke);

    //
    // The clues
    //
    solver.post(IntConstraintFactory.arithm(englishman,"=",red));
    solver.post(IntConstraintFactory.arithm(spaniard,"=",dog));
    solver.post(IntConstraintFactory.arithm(coffee,"=",green));
    solver.post(IntConstraintFactory.arithm(ukranian,"=",tea));
    solver.post(IntConstraintFactory.arithm(green,"=",VariableFactory.offset(ivory,1)));
    solver.post(IntConstraintFactory.arithm(old_gold,"=",snails));
    solver.post(IntConstraintFactory.arithm(kools,"=",yellow));
    solver.post(IntConstraintFactory.arithm(milk,"=",VariableFactory.fixed(3,solver)));
    solver.post(IntConstraintFactory.arithm(norwegian,"=",VariableFactory.fixed(1,solver)));
    solver.post(IntConstraintFactory.distance(fox, chesterfields,"=",1));
    solver.post(IntConstraintFactory.distance(horse, kools,"=",1));
    solver.post(IntConstraintFactory.arithm(lucky_strike,"=",fruit_juice));
    solver.post(IntConstraintFactory.arithm(japanese,"=",parliaments));
    solver.post(IntConstraintFactory.distance(norwegian, blue,"=",1));


  }


  @Override
  public void configureSearch() {
    solver.set(IntStrategyFactory.firstFail_InDomainMin(all));
  }
  
  @Override
  public void solve() {
    solver.findSolution();
  }

  @Override
  public void prettyOut() {

    if (solver.isFeasible() == ESat.TRUE) {
      int num_sol = 0;
      do {
        System.out.print("Colors      : ");
        for (int i = 0; i < n; i++) {
          System.out.print(colors[i].getValue() + " ");
        }
        System.out.print("\nNationality : ");
        for (int i = 0; i < n; i++) {
          System.out.print(nationality[i].getValue() + " ");
        }
        System.out.print("\nAnimal      : ");
        for (int i = 0; i < n; i++) {
          System.out.print(animal[i].getValue() + " ");
        }
        System.out.print("\nDrink       : ");
        for (int i = 0; i < n; i++) {
          System.out.print(drink[i].getValue() + " ");
        }
        System.out.print("\nSmoke       : ");
        for (int i = 0; i < n; i++) {
          System.out.print(smoke[i].getValue() + " ");
        }
        System.out.println();

        num_sol++;

      } while (solver.nextSolution() == Boolean.TRUE);
      
      System.out.println("\nIt was " + num_sol + " solutions.");
      
    } else {
      System.out.println("No solution.");
    }

  }

  public static void main(String[] args) {

    new Zebra().execute(args);

  }


}
 

File: z3/zebra.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Zebra puzzle in Z3
#
# Port of Google or-tools zebra.py:
# https://github.com/google/or-tools/blob/master/examples/python/zebra.py
# (it's not mine :-)
#
# """
# This is the zebra problem as invented by Lewis Caroll.
#
# There are five houses.
# The Englishman lives in the red house.
# The Spaniard owns the dog.
# Coffee is drunk in the green house.
# The Ukrainian drinks tea.
# The green house is immediately to the right of the ivory house.
# The Old Gold smoker owns snails.
# Kools are smoked in the yellow house.
# Milk is drunk in the middle house.
# The Norwegian lives in the first house.
# The man who smokes Chesterfields lives in the house next to the man
#    with the fox.
# Kools are smoked in the house next to the house where the horse is kept.
# The Lucky Strike smoker drinks orange juice.
# The Japanese smokes Parliaments.
# The Norwegian lives next to the blue house.
#
# Who owns a zebra and who drinks water?
# """
# 
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
#
from __future__ import print_function
from z3_utils_hakank import *


def main():

  sol = Solver()

  red = makeIntVar(sol, "red", 1, 5)
  green = makeIntVar(sol,"green", 1, 5)
  yellow = makeIntVar(sol,"yellow",1, 5)
  blue = makeIntVar(sol, "blue", 1, 5)
  ivory = makeIntVar(sol, "ivory", 1, 5)

  englishman = makeIntVar(sol, "enlighman", 1, 5)
  spaniard = makeIntVar(sol,"spaniard",1, 5)
  japanese = makeIntVar(sol,"japanese", 1, 5)
  ukrainian = makeIntVar(sol, "ukrainian", 1, 5)
  norwegian = makeIntVar(sol, "norwegian", 1, 5)

  dog = makeIntVar(sol, "dog", 1, 5)
  snails = makeIntVar(sol, "snails", 1, 5)
  fox = makeIntVar(sol, "fox", 1, 5)
  zebra = makeIntVar(sol, "zebra", 1, 5)
  horse = makeIntVar(sol, "horse", 1, 5)

  tea = makeIntVar(sol,"tea", 1, 5)
  coffee = makeIntVar(sol, "coffee", 1, 5)
  water = makeIntVar(sol, "water", 1, 5)
  milk = makeIntVar(sol, "milk", 1, 5)
  fruit_juice = makeIntVar(sol, "fruit juice", 1, 5)

  old_gold = makeIntVar(sol, "old gold", 1, 5)
  kools = makeIntVar(sol, "kools", 1, 5)
  chesterfields = makeIntVar(sol, "chesterfields", 1, 5)
  lucky_strike = makeIntVar(sol, "lucky strike", 1, 5)
  parliaments = makeIntVar(sol, "parliaments", 1, 5)

  sol.add(Distinct([red, green, yellow, blue, ivory]))
  sol.add(Distinct([englishman, spaniard, japanese,
                                  ukrainian, norwegian]))
  sol.add(Distinct([dog, snails, fox, zebra, horse]))
  sol.add(Distinct([tea, coffee, water, milk,
                                  fruit_juice]))
  sol.add(Distinct([parliaments, kools, chesterfields,
                                  lucky_strike, old_gold]))

  sol.add(englishman == red)
  sol.add(spaniard == dog)
  sol.add(coffee == green)
  sol.add(ukrainian == tea)
  sol.add(green == ivory + 1)
  sol.add(old_gold == snails)
  sol.add(kools == yellow)
  sol.add(milk == 3)
  sol.add(norwegian == 1)
  sol.add(Abs(fox - chesterfields) == 1)
  sol.add(Abs(horse - kools) == 1)
  sol.add(lucky_strike == fruit_juice)
  sol.add(japanese == parliaments)
  sol.add(Abs(norwegian - blue) == 1)

  all_vars = [parliaments, kools, chesterfields, lucky_strike, old_gold,
              englishman, spaniard, japanese, ukrainian, norwegian,
              dog, snails, fox, zebra, horse,
              tea, coffee, water, milk, fruit_juice,
              red, green, yellow, blue, ivory]

  if sol.check() == sat:
    mod = sol.model()
    people = [englishman, spaniard, japanese, ukrainian, norwegian]
    water_drinker = [p for p in people if mod.eval(p) == mod.eval(water)][0]
    zebra_owner = [p for p in people if mod.eval(p) == mod.eval(zebra)][0]
    print('The', water_drinker, 'drinks water.')
    print('The', zebra_owner, 'owns the zebra.')
  else:
    print('No solutions to the zebra problem, this is unusual!')



if __name__ == '__main__':
  main()
